### ISSUE TITLE
How to detect dropped TCP/SMPP connection

### ISSUE THREAD

**Issue by qosimmax** [linxGnu/gosmpp#31]
Hi @linxGnu 
When sending sms  periodically (every 10 seconds) ,  then i turned off internet connection(wi-fi).
There is no error when send sms.  How detect it?

**Comment by linxGnu** [comment]
Hello @qosimmax , please kindly take a look at following test/sample code: https://github.com/linxGnu/gosmpp/blob/master/transceiver_test.go#L53-L73

- You could detect submit failure or you could be notified when connection is closed
- `gosmpp` will try to rebind connection whenever it could and won't tell you about connection fail. But as I mentioned above, you could detect submit failure via https://github.com/linxGnu/gosmpp/blob/master/transceiver_test.go#L56, it's your own callback to handle submit failure.

**Comment by qosimmax** [comment]
Yes i am using  callback handler OnSubmitError ,  but this handler not working when i turned off internet.

**Comment by linxGnu** [comment]
@qosimmax Oh it's weird, let me take a look later and tell you asap :thinking:

**Comment by linxGnu** [comment]
@qosimmax 
I reran example at: https://github.com/linxGnu/gosmpp/tree/master/example

SMPP Client:
```go
go run main.go
```

SMSC (at https://github.com/linxGnu/gosmpp/tree/master/example/smsc):
```bash
# compile and run smsc
sh run.sh
```

Sometime turn on/off SMSC, I saw following error raised from smpp client:
```
dial tcp 127.0.0.1:2775: connect: connection refused
Transmitter is closing. Can not send PDU to SMSC
Transmitter is closing. Can not send PDU to SMSC
```

In short:
- SubmitError callbacked when you submit to SMSC but fail. I think in your case, SubmitSM success before turn off internet
- You could check RebindingError callback. Since smpp client retry connect to SMSC automatically, when you turn off internet, it tries to rebind but failure.

Could you please check it again, maybe using RebindingError callback as the example

**Comment by qosimmax** [comment]
@linxGnu  you run SMSC in the localhost.
When run smpp client  and SMSC in same host (localhost) its work fine,  problem is when connect to remote SMSC host.
Can you test smmp client with remote host(SMSC simulator) ? 
https://melroselabs.com/services/smsc-simulator/

**Comment by linxGnu** [comment]
@qosimmax I could reproduce your concern and provide an update. Please kindly check the fix again with latest rc release.

tl;dr: it's go problem of detecting conn broken without specifying read/write deadline.

Thank you very much for using gosmpp and tell us your concern/issue.

**Comment by qosimmax** [comment]
thanks very much @linxGnu.
If i'm not mistaken  value of ReadTimeout must greater than EnquireLinkIntervalMinimum (20 second),
https://github.com/linxGnu/gosmpp/blob/297995d40f7e6fc4d8926092a7e32f63e334f089/transmitter.go#L15
otherwise EnquireLinkResp handle not working.

**Comment by linxGnu** [comment]
@qosimmax Actually, we can't tell when `EnquireLinkResp` is returned from SMSC. Therefore, I don't think there is a mandatory condition like `ReadTimeout >= EnquireLinkInterval`. 

So, I think I would like to give the choice to user. User should try with their own production environment

**Comment by linxGnu** [comment]
I will close this issue. Thank you very much for your concern @qosimmax
